### PR TITLE
compilers/cpp: check libc++ vs libstdc++ harder

### PR DESCRIPTION
**Pull request by dcbaker**
Instead of hardcoding any values, hardcode what we think the most likely implementation is, and check that first. It was pointed out that while for example, Apple only provides libc++ and supports that for xcode, a user could install a custom environment (such as homebrew) which uses it's own copy of libstdc++, and we need to account for that. This means that a library search will be done, but only once and the result will be cached, on all systems.

### CODE REVIEW

**Comment by eli-schwartz** [comment]
> Instead of hardcoding any values, hardcode what we think the most likely implementation is, and check that first.

So if I understand this correctly, this still always produces libc++ on Apple etc. systems, and libstdc++ on Linux etc. systems, unless the compiler doesn't support the OS default at all. If it supports both, we always use the OS default even if it is the opposite of the current compiler's default.

Previously, if it was the OS default we hardcoded that.

So the difference in behavior is now for the case where:
1) you're on a system where the default is libc++
2) you're using a non-default toolchain
3) that toolchain doesn't support the OS default even as an opt-in

... and before this commit that would just totally break, I guess.

Am I understanding the problem we're focused on solving here?

**Comment by eli-schwartz** [comment]
> The only way I can think of to fix that case is to add an option like `cpp_stdlib=c++ | stdc++`, to allow a user to force a choice.

Or detect `-stdlib=lib{std}c++` in the cpp_args+link_args / CXXFLAGS + LDFLAGS? This does feel suboptimal, I do admit.

**Comment by dcbaker** [comment]
The great part about that is that some clang's silently ignore/rewrite `-stdlib=libstdc++` into `-stdlib=libc++`. Apple's clang does that, and I think I've seen that the android clang does that too

File: mesonbuild/compilers/cpp.py [hunk 1]
```diff
@@ -206,22 +206,19 @@ def language_stdlib_only_link_flags(self, env: Environment) -> T.List[str]:
         machine = env.machines[self.for_machine]
         assert machine is not None, 'for mypy'
 
-        # We need to determine whether to us libc++ or libstdc++ In some cases
-        # we know the answer, so we'll hardcode those cases.  There are other
-        # cases where we can't know the answer just by looking at the OS, namely
-        # on Linux. In that case we have to fallback to manually checking
-        stdlib: str
+        # We need to determine whether to us libc++ or libstdc++. We can't
```
Comment (tristan957) [inline mesonbuild/compilers/cpp.py:209]: s/us/use

**Comment by dcbaker** [comment]
> Or detect -stdlib=lib{std}c++ in the cpp_args+link_args / CXXFLAGS + LDFLAGS? This does feel suboptimal, I do admit.

In general we don't do this, except to warn you not to do it. I feel like this is roughly equivalent to Windows MSCRT options (mtd, mt, md, mdd). So the more I think about it the more I think I'd prefer to handle it with a builtin option.

**Comment by barracuda156** [comment]
@eli-schwartz @dcbaker 
> The problem that still exists is that if you have the platform default in your search path, and you want to use a different one, that will fail.

It does indeed. I wasted a couple of hours now trying to figure out who smuggles in `-lc++` flag for Qemu build, where compiler is GCC (should already exclude `libc++` from defaults at least) and the arch is `ppc` (where `libc++` was never supported on MacOS and modern versions won’t even build). Well, looks like it is Meson, since instead of borrowing sane settings from the compiler or environment, it apparently searches through directories – and lo and behold, I had `libc++.dylib` – an experimental build which never posed a problem since GCC is smart enough to link to its own libraries. Here this mechanism is broken, and choice is forced upon. The same problem will arise on Intel, I believe, where you cannot even get rid of `libc++`, it is a part of the OS.

**Pull request merged**